ihmc-communication/src/main/java/us/ihmc/communication/ros2/sync/ROS2PeerClockOffsetEstimator.java

```diff
@@ -0,0 +1,143 @@
+
+         if (receivedMessage.getIsRequest() && receivedRequestTarget.equals(ourGuid)) // Reply
+         {
+            PeerClockOffsetEstimatorPingMessage replyMessage = new PeerClockOffsetEstimatorPingMessage();
+            replyMessage.set(receivedMessage);
+            replyMessage.setIsRequest(false);
+            MessageTools.toMessage(Instant.now(), replyMessage.getReplySendTime());
+
+            ThreadTools.startAsDaemon(() ->
+            {
+               synchronized (publisher)
+               {
+                  publisher.publish(replyMessage);
+               }
+            }, getClass().getSimpleName() + "PublishReply");
+         }
+         else if (!receivedMessage.getIsRequest() && receivedReplyTarget.equals(ourGuid)) // Update clock offset estimate
+         {
+            ROS2PeerClockOffsetEstimatorPeer peer = peerMap.get(receivedRequestTarget);
+            if (peer != null)
+            {
+               peer.update(MessageTools.toInstant(receivedMessage.getRequestSendTime()),
+                           Instant.now(),
+                           MessageTools.toInstant(receivedMessage.getReplySendTime()));
+            }
+         }
+      }, (subscriber, info) ->
+      {
+         Guid guid = info.getGuid();
+         if (!guid.equals(publisher.getPublisher().getGuid())) // Exclude our publisher
+         {
+            switch (info.getStatus())
+            {
+               case MATCHED_MATCHING ->
+               {
+                  if (!peerMap.containsKey(guid))
+                  {
+                     Guid guidCopy = new Guid();
+                     guidCopy.set(guid);
+
+                     ROS2PeerClockOffsetEstimatorPeer peer = new ROS2PeerClockOffsetEstimatorPeer(guidCopy);
+                     peerMap.put(guidCopy, peer);
+                     peerList.add(peer);
+                  }
+               }
+               case REMOVED_MATCHING ->
+               {
+                  peerMap.remove(guid);
+                  peerList.removeIf(peer -> peer.getGuid().equals(guid));
+               }
+            }
+         }
+      });
+
+      requestThread.setFrequencyLimit(5.0);
+      requestThread.startRepeating();
+   }
+
+   private void requestThread()
+   {
+      if (!peerList.isEmpty())
+      {
+         if (nextPeerToPing >= peerList.size())
+            nextPeerToPing = 0;
+
+         ROS2PeerClockOffsetEstimatorPeer peer = peerList.get(nextPeerToPing);
+         requestMessage.setIsRequest(true);
+         MessageTools.toMessage(peer.getGuid(), requestMessage.getRequestTarget());
+         MessageTools.toMessage(ourGuid, requestMessage.getReplyTarget());
+         MessageTools.toMessage(Instant.now(), requestMessage.getRequestSendTime());
+         synchronized (publisher)
+         {
+            publisher.publish(requestMessage);
+         }
+
+         ++nextPeerToPing;
+      }
+   }
+
+   public void destroy()
```

Comment:
Should this also call `remove()` on the subscriber and publisher? I never really understood if that's necessary or not

Reply:
I can do that. When you destroy the parent node, it removes all the pubs/subs automatically, but it doesn't hurt to remove them early.